diff --git a/arch/x86/entry/syscalls/syscall_64.tbl b/arch/x86/entry/syscalls/syscall_64.tbl
index c29976eca..f5df3b093 100644
--- a/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/arch/x86/entry/syscalls/syscall_64.tbl
@@ -357,6 +357,7 @@
 433	common	fspick			__x64_sys_fspick
 434	common	pidfd_open		__x64_sys_pidfd_open
 435	common	clone3			__x64_sys_clone3/ptregs
+449	common  ptree			sys_ptree
 
 #
 # x32-specific system call numbers start at 512 to avoid cache impact
diff --git a/include/linux/syscalls.h b/include/linux/syscalls.h
index f7c561c4d..c01da1849 100644
--- a/include/linux/syscalls.h
+++ b/include/linux/syscalls.h
@@ -69,6 +69,7 @@ struct rseq;
 union bpf_attr;
 struct io_uring_params;
 struct clone_args;
+struct prinfo;
 
 #include <linux/types.h>
 #include <linux/aio_abi.h>
@@ -648,6 +649,8 @@ asmlinkage long sys_syslog(int type, char __user *buf, int len);
 /* kernel/ptrace.c */
 asmlinkage long sys_ptrace(long request, long pid, unsigned long addr,
 			   unsigned long data);
+
+asmlinkage int sys_ptree(struct prinfo *buf, int *nr, int pid);
 /* kernel/sched/core.c */
 
 asmlinkage long sys_sched_setparam(pid_t pid,
diff --git a/kernel/Makefile b/kernel/Makefile
index daad787fb..2088d1ce3 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -10,7 +10,7 @@ obj-y     = fork.o exec_domain.o panic.o \
 	    extable.o params.o \
 	    kthread.o sys_ni.o nsproxy.o \
 	    notifier.o ksysfs.o cred.o reboot.o \
-	    async.o range.o smpboot.o ucount.o
+	    async.o range.o smpboot.o ucount.o ptree.o
 
 obj-$(CONFIG_MODULES) += kmod.o
 obj-$(CONFIG_MULTIUSER) += groups.o
diff --git a/kernel/ptree.c b/kernel/ptree.c
index 0b23f853a..eef62efa3 100644
--- a/kernel/ptree.c
+++ b/kernel/ptree.c
@@ -24,7 +24,7 @@ void unregister_ptree(ptree_func func)
 asmlinkage int sys_ptree(struct prinfo *buf, int *nr, int pid)
 {
 	int rc = -1;
-	int _nr = NULL;
+	int _nr = 0;
 	int bytes_copied = 0;
 	struct prinfo *_buf = NULL;
 
@@ -45,7 +45,7 @@ asmlinkage int sys_ptree(struct prinfo *buf, int *nr, int pid)
 		goto Exit;
 	}
 
-	_buf = (struct prinfo*)kmalloc(_nr * sizeof(struct prinfo));
+	_buf = (struct prinfo *)kmalloc(_nr * sizeof(struct prinfo), GFP_KERNEL);
 	if (_buf == NULL) {
 		pr_crit("Failed to kmalloc\n");
 		rc = -ENOSYS;
diff --git a/kernel/ptree.h b/kernel/ptree.h
index 5182b0f08..bbbfe5bb7 100644
--- a/kernel/ptree.h
+++ b/kernel/ptree.h
@@ -1,8 +1,10 @@
-#include <include/uapi/asm-generic/errno-base.h>
 #include <linux/kernel.h>
-#include <include/linux/types.h>
-#include <include/linux/kmod.h>
-#include <include/linux/printk.h>
+#include <linux/types.h>
+#include <linux/kmod.h>
+#include <linux/printk.h>
+#include <linux/ptrace.h>
+#include <linux/slab.h>
+
 
 
 struct prinfo {
@@ -15,5 +17,6 @@ struct prinfo {
     };
 
 typedef int (*ptree_func)(struct prinfo *buf, int *nr, int pid);
+asmlinkage int sys_ptree(struct prinfo *buf, int *nr, int pid);
 int register_ptree(ptree_func func);
 void unregister_ptree(ptree_func func);
